internal/app/synchronizer/synchronizer_service.go

```diff
@@ -91,67 +93,85 @@ func (s *synchronizerService) Synchronize(stream ipb.Synchronizer_SynchronizeSer
 	// 1. Receive proposals from backend, send them to cycle.
 	// 2. Receive matches and signals from cycle, send them to backend.
 
-	registration := s.register(stream.Context())
+	eg, ctx := errgroup.WithContext(stream.Context())
+	registration := s.register(ctx)
 	m6cBuffer := bufferStringChannel(registration.m7c)
 	defer func() {
 		for range m6cBuffer {
 		}
 	}()
 
-	go func() {
+	matches := map[string]*pb.Match{}
+	eg.Go(func() error {
+		defer registration.allM1cSent.Done()
+
+		duplicateIDs := []string{}
 		for {
 			req, err := stream.Recv()
+			if err == io.EOF {
+				break
+			}
 			if err != nil {
-				if err != io.EOF {
-					logger.WithFields(logrus.Fields{
-						"error": err.Error(),
-					}).Error("error streaming in synchronizer from backend")
-				}
-				registration.allM1cSent.Done()
-				return
+				return fmt.Errorf("error streaming in synchronizer from backend, desc: %w", err)
 			}
-			registration.m1c.send(mAndM6c{m: req.Proposal, m7c: registration.m7c})
+			if _, ok := matches[req.GetProposal().GetMatchId()]; ok {
+				duplicateIDs = append(duplicateIDs, req.GetProposal().GetMatchId())
+			}
+			matches[req.GetProposal().GetMatchId()] = req.GetProposal()
+		}
+
+		if len(duplicateIDs) != 0 {
```

Comment:
The synchronize method only has context from one mmf run... this isn't checking different mmf calls....

See comment on evaluator client for alternative.